pkg/natdiscovery/remote_endpoint.go

```diff
@@ -0,0 +1,138 @@
+	publicToPrivateFailoverTimeout = 1 * time.Second
+)
+
+type remoteEndpointNAT struct {
+	Endpoint               *types.SubmarinerEndpoint
+	state                  endpointState
+	lastCheck              time.Time
+	lastTransition         time.Time
+	started                time.Time
+	UseNAT                 bool
+	UseIP                  string
+	lastPublicIPRequestID  uint64
+	lastPrivateIPRequestID uint64
+}
+
+func newRemoteEndpointNAT(endpoint *types.SubmarinerEndpoint) *remoteEndpointNAT {
+	return &remoteEndpointNAT{
+		Endpoint:       endpoint,
+		state:          testingPrivateAndPublicIPs,
+		started:        time.Now(),
+		lastTransition: time.Now(),
+	}
+}
+
+func (rnat *remoteEndpointNAT) transitionToState(newState endpointState) {
+	rnat.lastTransition = time.Now()
+	rnat.state = newState
+}
+
+func (rnat *remoteEndpointNAT) sinceLastTransition() time.Duration {
+	return time.Since(rnat.lastTransition)
+}
+
+func (rnat *remoteEndpointNAT) hasTimedOut() bool {
+	return time.Since(rnat.started) > totalTimeout
+}
+
+func (rnat *remoteEndpointNAT) useLegacyNATSettings() {
+	rnat.UseNAT = rnat.Endpoint.Spec.NATEnabled
+	if rnat.Endpoint.Spec.NATEnabled {
+		rnat.UseIP = rnat.Endpoint.Spec.PublicIP
+		rnat.transitionToState(selectedPublicIP)
+	} else {
+		rnat.UseIP = rnat.Endpoint.Spec.PrivateIP
+		rnat.transitionToState(selectedPrivateIP)
+	}
+}
+
+func (rnat *remoteEndpointNAT) shouldCheck() bool {
+	switch rnat.state {
+	case testingPrivateAndPublicIPs:
+		return true
+	case waitingForResponse:
+		return time.Since(rnat.lastCheck) > recheckTime
+	default:
+		return false
+	}
+}
+
+func (rnat *remoteEndpointNAT) checkSent() {
+	if rnat.state == testingPrivateAndPublicIPs {
+		rnat.state = waitingForResponse
+	}
+
+	rnat.lastCheck = time.Now()
+}
+
+func (rnat *remoteEndpointNAT) transitionToPublicIP(remoteEndpointID string, useNAT bool) error {
+	if rnat.state == waitingForResponse {
+		rnat.UseIP = rnat.Endpoint.Spec.PublicIP
+		rnat.transitionToState(selectedPublicIP)
+		rnat.UseNAT = useNAT
+	} else {
+		return errors.Errorf("received unexpected transition to public IP from endpoint %q", remoteEndpointID)
+	}
+
+	return nil
+}
+
+func (rnat *remoteEndpointNAT) transitionToPrivateIP(remoteEndpointID string, useNAT bool) error {
+	switch rnat.state {
+	case waitingForResponse:
+		rnat.UseIP = rnat.Endpoint.Spec.PrivateIP
+		rnat.transitionToState(selectedPrivateIP)
+
+	case selectedPublicIP:
+		// If a PublicIP was selected, we still allow some time for the privateIP response to arrive, and we always
+		// prefer PrivateIP with no NAT connection, as it will be more likely to work, and more efficient
+		if rnat.sinceLastTransition() > publicToPrivateFailoverTimeout {
+			return errors.Errorf("too late response on private address received after response on public address for endpoint %q",
+				remoteEndpointID)
+		}
+
+		rnat.UseIP = rnat.Endpoint.Spec.PrivateIP
+		rnat.transitionToState(selectedPrivateIP)
+	case selectedPrivateIP:
```

Comment:
I think @mangelajo wants default to be executed only when rnat.state is NOT `selectedPrivateIP` so it might be needed.

Reply:
correct, we need it for what @sridhargaddam said.

It looks weird, because in C that meant that selectedPrivateIP: == default: but in go it's just no action for selectedPrivateIP.

And it's a case which could happen (for example if both public & private respond) I can add at least a warning message to make it clearer.

Reply:
ah right - I was thinking fall-through